ntp-daemon/src/ipfilter.rs

```diff
@@ -0,0 +1,328 @@
+#[derive(Debug, Copy, Clone, Default, PartialEq, Eq)]
+struct Treenode {
+    // Where in the array the child nodes of this
+    // node are located. A child node is only
+    // generated if the symbol cannot be used to
+    // make a final decision at this level
+    child_offset: u32,
+    inset: u16,
+    outset: u16,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// BitTree is a Trie on 128 bit integers encoding
+/// which integers are part of the set.
+///
+/// It matches the integer a 4-bit segment at a time
+/// recording at each level whether for a given symbol
+/// all integers with the prefix extended with that
+/// symbol are either in or outside of the set.
+struct BitTree {
+    nodes: Vec<Treenode>,
+}
+
+fn top_quad(v: u128) -> u8 {
+    ((v & (0xF << 124)) >> 124) as u8
+}
+
+impl BitTree {
+    #[allow(dead_code)]
+    /// Lookup whether a given value is in the set encoded in this BitTree
+    /// Complexity is O(log(l)), where l is the length of the longest
+    /// prefix in the set.
+    fn lookup(&self, mut val: u128) -> bool {
+        let mut node = &self.nodes[0];
+        loop {
+            // extract the current symbol as bit and see if we know the answer immediately.
+            // (example: symbol 1 maps to 0x2, symbol 5 maps to 0x10)
+            let cur = 1 << top_quad(val);
+            if node.inset & cur != 0 {
+                return true;
+            }
+            if node.outset & cur != 0 {
+                return false;
+            }
+            // no decision, shift to next symbol
+            val <<= 4;
+            // To calculate the child index we need to know how many symbols smaller
+            // than our symbol are not decided here. We do this by generating the bitmap
+            // of symbols neither in in or out, then masking out all symbols >=cur
+            // and finaly counting how many are left.
+            let next_idx =
+                node.child_offset + (!(node.inset | node.outset) & (cur - 1)).count_ones();
+            node = &self.nodes[next_idx as usize];
+        }
+    }
+
+    #[allow(dead_code)]
+    /// Create a BitTree from the given prefixes. Complexity is O(n*log(l)),
+    /// where n is the number of prefixes, and l the length of the longest
+    /// prefix.
+    fn create(data: &mut [(u128, u8)]) -> Self {
+        // Ensure values only have 1s in significant positions
+        for (val, len) in data.iter_mut() {
+            *val &= 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF_u128
+                .checked_shl((128 - *len) as u32)
+                .unwrap_or(0);
+        }
+        // Ensure values are sorted by value and then by length
+        data.sort_by(
+            |(v1, l1), (v2, l2)| {
+                if *v1 != *v2 {
+                    v1.cmp(v2)
+                } else {
+                    l1.cmp(l2)
+                }
+            },
+        );
+
+        let mut result = BitTree {
+            nodes: [Treenode::default()].into(),
```

Comment:
eh, why not `vec![Treenode::default()]`?

Reply:
fixed